crates/phactory/src/prpc_service.rs

```diff
@@ -1807,6 +1839,213 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
+        &mut self,
+        _request: (),
+    ) -> RpcResult<pb::DcapHandoverChallenge> {
+        let ntp_time_secs = crate::nts::nts_get_time_secs()
+            .await
+            .map_err(from_display)?;
+        let mut phactory = self.lock_phactory(false, true)?;
+        if phactory.attestation_provider != Some(AttestationProvider::Dcap) {
+            return Err(from_display("DCAP not enabled"));
+        }
+        let (block, block_time_ms) = phactory.current_block()?;
+        let a_week = 7 * 24 * 60 * 60;
+        if block_time_ms / 1000 + a_week < ntp_time_secs {
+            return Err(from_display(
+                "Current block is too old, please sync to latest",
+            ));
+        }
+        let challenge = phactory.get_dcap_worker_key_challenge(block, block_time_ms, ntp_time_secs);
+        Ok(pb::DcapHandoverChallenge::new(challenge))
+    }
+
+    async fn dcap_handover_start(
+        &mut self,
+        request: pb::DcapHandoverChallengeResponse,
+    ) -> RpcResult<pb::DcapHandoverWorkerKey> {
+        let ntp_now = crate::nts::nts_get_time_secs()
+            .await
+            .map_err(from_display)?;
+        let mut phactory = self.lock_phactory(false, true)?;
+        if phactory.attestation_provider != Some(AttestationProvider::Dcap) {
+            return Err(from_display("DCAP not enabled"));
+        }
+        let dev_mode = phactory.dev_mode;
+        let (block_number, _) = phactory.current_block()?;
+
+        let challenge_handler = request.decode_challenge_handler().map_err(from_display)?;
+
+        // 1. verify challenge validity to prevent replay attack
+        let challenge = &challenge_handler.challenge;
+        if !phactory.dcap_verify_worker_key_challenge(challenge) {
+            return Err(from_display("Invalid challenge"));
+        }
+        // 2. ensure delta time between client and server is within 1 minutes
+        if challenge.ntp_time_secs > ntp_now || ntp_now - challenge.ntp_time_secs > 60 {
+            return Err(from_display("Invalid NTP time"));
+        }
+        // 3. verify sgx local attestation report to ensure the handover pRuntimes are on the same machine
+        let recv_local_report = unsafe {
+            sgx_api_lite::decode(&request.la_report)
+                .map_err(|_| from_display("Invalid client LA report"))?
+        };
+        sgx_api_lite::verify(recv_local_report).map_err(|_| from_display("No remote handover"))?;
+        let handler_hash = blake2_256(&request.encoded_challenge_handler);
+        if handler_hash != recv_local_report.body.report_data[..32] {
+            return Err(from_display("Invalid challenge handler"));
+        }
+        // 4. verify challenge block height and report timestamp
+        // only challenge within 150 blocks (30 minutes) is accepted
+        let challenge_height = challenge.block_number;
+        if !(challenge_height <= block_number && block_number - challenge_height <= 150) {
```

Comment:
This seems superceded by the `challenge.ntp_time_secs` check above

Reply:
Removed this one.